common/setups/rasr/nn_system.py

```diff
@@ -0,0 +1,554 @@
+        feature_flow.interconnect(
+            base_flow, base_mapping, tf_flow, tf_mapping, {base_output: input_name}
+        )
+        feature_flow.interconnect_outputs(tf_flow, tf_mapping)
+
+        return feature_flow
+
+    # -------------------- Setup --------------------
+    def init_system(
+        self,
+        hybrid_init_args: RasrInitArgs,
+        train_data: Dict[str, Union[ReturnnRasrDataInput, OggZipHdfDataInput]],
+        cv_data: Dict[str, Union[ReturnnRasrDataInput, OggZipHdfDataInput]],
+        devtrain_data: Optional[
+            Dict[str, Union[ReturnnRasrDataInput, OggZipHdfDataInput]]
+        ] = None,
+        dev_data: Optional[Dict[str, ReturnnRasrDataInput]] = None,
+        test_data: Optional[Dict[str, ReturnnRasrDataInput]] = None,
+        train_cv_pairing: Optional[List[Tuple[str]]] = None,
+    ):
+        self.hybrid_init_args = hybrid_init_args
+
+        self._init_am(**self.hybrid_init_args.am_args)
+
+        devtrain_data = devtrain_data if devtrain_data is not None else {}
+        dev_data = dev_data if dev_data is not None else {}
+        test_data = test_data if test_data is not None else {}
+
+        self._assert_corpus_name_unique(
+            train_data, devtrain_data, cv_data, dev_data, test_data
+        )
+
+        self.train_input_data = train_data
+        self.cv_input_data = cv_data
+        self.devtrain_input_data = devtrain_data
+        self.dev_input_data = dev_data
+        self.test_input_data = test_data
+
+        self.train_corpora.extend(list(train_data.keys()))
+        self.cv_corpora.extend(list(cv_data.keys()))
+        self.devtrain_corpora.extend(list(devtrain_data.keys()))
+        self.dev_corpora.extend(list(dev_data.keys()))
+        self.test_corpora.extend(list(test_data.keys()))
+
+        self.train_cv_pairing = train_cv_pairing
+
+        for trn_c, cv_c in itertools.product(self.train_corpora, self.cv_corpora):
+            if (trn_c, cv_c) in self.train_cv_pairing or self.train_cv_pairing is None:
+                continue
+
+            self.jobs[f"{trn_c}_{cv_c}"] = {}
+            self.nn_models[f"{trn_c}_{cv_c}"] = {}
+            self.nn_configs[f"{trn_c}_{cv_c}"] = {}
+
+    def prepare_data(self, raw_sampling_rate: int, feature_sampling_rate: int):
+        for name in self.train_corpora + self.devtrain_corpora + self.cv_corpora:
+            self.jobs[name]["ogg_zip"] = j = returnn.BlissToOggZipJob(
+                bliss_corpus=self.crp[name].corpus_config.corpus_file,
+                segments=self.crp[name].segment_path,
+                rasr_cache=self.feature_flows[name]["init"],
+                raw_sample_rate=raw_sampling_rate,
+                feat_sample_rate=feature_sampling_rate,
+            )
+            self.oggzips[name] = j.out_ogg_zip
+            j.add_alias(f"oggzip/{name}")
+
+            # TODO self.jobs[name]["hdf_full"] = j = returnn.ReturnnDumpHDFJob()
+
+    def generate_lattices(self):
+        pass
```

Comment:
technically we could even do sequence training for GMMs. Just nobody since Zoltan has done this and I don't even have an example config for this.

So while in general this function could go into the `RasrSystem`, we can also leave it here.

Reply:
I can do a PR to add the lattices if you want. Just remind me some time ;)

Reply:
Ah then I would prefer to leave this function here :)

> I can do a PR to add the lattices if you want. Just remind me some time ;)

yes please :)